x-pack/plugins/alerting/server/task_runner/task_runner.ts

```diff
@@ -701,6 +708,18 @@ export class TaskRunner<
     ruleMonitoring.execution.calculated_metrics.success_ratio =
       ruleMonitoring.execution.history.filter(({ success }) => success).length /
       ruleMonitoring.execution.history.length;
+      
+    // Percentile calculations for rule executions, limited to predefined number of samples
+    const historyWithDuration = ruleMonitoring.execution.history.filter(history => history.duration);
+    // Reformat history array to numbers array (durations in ms)
+    const durationSamples = historyWithDuration.map(history => history.duration);
+    const slicedDurationSamples = durationSamples.slice(-MONITORING_PERCENTILE_SAMPLE_LIMIT);
+    
+    // Calculating P50/95/99, empty arrays result in NaN, so defaulting to 0
+    ruleMonitoring.execution.calculated_metrics.p50 = stats.percentile(slicedDurationSamples, 0.50) || 0;
+    ruleMonitoring.execution.calculated_metrics.p95 = stats.percentile(slicedDurationSamples, 0.95) || 0;
+    ruleMonitoring.execution.calculated_metrics.p99 = stats.percentile(slicedDurationSamples, 0.99) || 0;
```

Comment:
Question: How exactly do we want to denote rules without a P50/95/99 available? It seems like the `lastDuration` is a nullable field, also we might not get enough data to calculate a percentile initially.

The other question is what do we want to display in the UX if there is no precentiles (0ms?, something like N/A)?

Here I have the fields default to `0` if the above 2 cases are true. In the UX it currently displays `0ms`.

Open to suggestions :smiley: